Topics/Tech_Stacks/WebSockets/Web-Sockets.md

```diff
@@ -0,0 +1,169 @@
+
+### How the WebSocket Connection is Established:
+1. **Client Handshake Request:** The client initiates the connection by sending a WebSocket handshake request to the server. This request looks like a standard HTTP request but includes an `Upgrade: websocket` header that signals the desire to establish a WebSocket connection.
+   
+2. **Server Handshake Response:** If the server supports WebSockets and accepts the connection request, it responds with a handshake of its own, confirming the upgrade from HTTP to WebSockets.
+   
+3. **Persistent Connection:** With the handshake complete, the connection remains open, and data can now be sent in either direction at any time, allowing for two-way, **real-time communication**.
+
+### WebSocket Communication:
+Once established, the WebSocket connection allows messages to be passed back and forth between the client and server with minimal overhead. Each message is framed in a standardized format, which can be quickly parsed and processed by both parties. This efficient messaging system is key to the protocol's ability to support high-frequency, real-time communication scenarios.
+
+WebSockets are particularly adept at handling use cases where the server needs to send data to the client proactively, without the client first making a request. This capability has led to the widespread adoption of WebSockets in various applications that require real-time functionality, such as live notifications, instant messaging, and collaborative editing platforms.
+
+In the context of CSC301, understanding WebSockets is essential for building modern applications where real-time user interaction is a core requirement. With WebSockets, software engineers can create responsive and interactive user experiences that were once difficult to achieve with traditional web technologies.
+
+![cpVnative](http_vs_websocket.png)
+
+## WebSockets vs HTTP
+HTTP, a foundational technology of the web, is designed around a stateless request-response model. This means that for each piece of data exchanged between the client and server, a new connection is initiated, followed by a request from the client and a response from the server. While effective for many web applications, this approach can introduce latency and inefficiency, especially in scenarios that demand real-time updates.
+
+In contrast, WebSockets offer a more dynamic communication protocol. After completing an initial handshake, WebSockets establish a stateful, two-way communication channel. This persistent connection remains open, allowing for continuous data exchange without the need to establish new connections for each interaction. This fundamental difference in operation between WebSockets and HTTP brings several notable advantages in specific contexts:
+
+- **Reduced Latency:** By eliminating the need for repeated HTTP handshakes, WebSockets significantly decrease latency, making them ideal for time-sensitive applications.
```

Comment:
Include examples of specific scenarios where reduced latency and real-time data flow are critical, such as online gaming or live auction bidding.

Reply:
Noted.